File: src/my_ldap_subjects.rs
use std::collections::{BTreeMap, HashMap};

use crate::my_types::*;
use crate::my_err::{Result};
use crate::ldap_wrapper::{LdapW};
use crate::my_ldap::{dn_to_rdn_and_parent_dn};
use crate::my_ldap::{url_to_dn_};
use crate::ldap_filter;


impl SubjectSourceConfig {
    pub fn search_filter_(&self, term: &str) -> String {
        self.search_filter.replace("%TERM%", term).replace(" ", "")
    }
}

async fn get_subjects_from_same_branch(ldp: &mut LdapW<'_>, sscfg: &SubjectSourceConfig, base_dn: &Dn, rdns: &[&str], search_token: &Option<String>) -> Result<Subjects> {
    let rdns_filter = ldap_filter::or(rdns.iter().map(|rdn| ldap_filter::rdn(rdn)).collect());
    let filter = if let Some(term) = search_token {
        ldap_filter::and2(&rdns_filter,&sscfg.search_filter_(term))
    } else {
        rdns_filter
    };
    Ok(ldp.search_subjects(base_dn, &sscfg.display_attrs, dbg!(&filter), None).await?)
}

pub async fn get_subjects_from_urls(ldp: &mut LdapW<'_>, urls: Vec<String>) -> Result<Subjects> {
    get_subjects(ldp, urls.into_iter().filter_map(url_to_dn_).collect(), &None, &None).await
}

fn into_group_map<K: Eq + std::hash::Hash, V, I: Iterator<Item = (K, V)>>(iter: I) -> HashMap<K, Vec<V>> {
    iter.fold(HashMap::new(), |mut map, (k, v)| {
        map.entry(k).or_insert_with(|| Vec::new()).push(v);
        map
    })
}
pub async fn get_subjects(ldp: &mut LdapW<'_>, dns: Vec<Dn>, search_token: &Option<String>, sizelimit: &Option<usize>) -> Result<Subjects> {
    let mut r = BTreeMap::new();


    let parent_dn_to_rdns = into_group_map(dns.iter().filter_map(|dn| {
        let (rdn, parent_dn) = dn_to_rdn_and_parent_dn(dn)?;
        Some((parent_dn, rdn))
    }));
        
    for (parent_dn, rdns) in parent_dn_to_rdns {
        let parent_dn = Dn::from(parent_dn);
        if let Some(sscfg) = ldp.config.dn_to_subject_source_cfg(&parent_dn) {
            let mut count = 0;
            for rdns_ in rdns.chunks(10) {
                let subjects = &mut get_subjects_from_same_branch(ldp, sscfg, &parent_dn, rdns_, search_token).await?;
                count += subjects.len();
                r.append(subjects);
                if let Some(limit) = sizelimit {
                    if count >= *limit { break; }
                }
            }
        }
    }

    Ok(r)
}

File: server/my_ldap_subjects.ts
import _ from "lodash";
import ldap_filter from "./ldap_filter";
import * as ldp from "./ldap_read_search"
import { dn_to_rdn_and_parent_dn, dn_to_sgroup_id, dn_to_subject_source_cfg, urls_to_dns } from "./dn";
import { Dn, DnsOpts, hMyMap, MyMap, Option, SubjectAttrs, Subjects, SubjectSourceConfig, toDn } from "./my_types";
import { mono_attrs } from "./ldap_helpers";
import { get_delete } from "./helpers";

export const hSubjectSourceConfig = {
    export: (self: SubjectSourceConfig) => (
        _.omit(self, ['search_filter'])
    ),
    search_filter_: (sscfg: SubjectSourceConfig, term: string) => (
        sscfg.search_filter.replace(/%TERM%/g, term).replace(/ /g, "")
    ),
}

export async function search_subjects(base_dn: Dn, attrs: string[], filter: string, dn2opts: DnsOpts, sizeLimit: Option<number>): Promise<Subjects> {
    const entries = await ldp.searchRaw(base_dn, filter, attrs, { sizeLimit });
    return _.fromPairs(entries.map(entry => { 
        const sgroup_id = dn_to_sgroup_id(entry.dn);
        const options = get_delete(dn2opts, entry.dn) ?? {};
        const subjectAttrs: SubjectAttrs = { attrs: mono_attrs(entry), sgroup_id, options };
        return [ toDn(entry.dn), subjectAttrs ]
    }))
}   

async function get_subjects_from_same_branch(sscfg: SubjectSourceConfig, base_dn: Dn, rdns: string[], dn2opts: DnsOpts, search_token: Option<string>) {
    const rdns_filter = ldap_filter.or(rdns.map(ldap_filter.rdn));
    const filter = search_token ?
        ldap_filter.and2(rdns_filter, hSubjectSourceConfig.search_filter_(sscfg, search_token)) :
        rdns_filter
    return await search_subjects(base_dn, sscfg.display_attrs, filter, dn2opts, undefined)
}

export const get_subjects_from_urls = async (urls: string[]) => (
    await get_subjects_(urls_to_dns(urls))
)

const fromPairsGrouped = <K extends string, V>(l: [K,V][]): MyMap<K, V[]> => (
    hMyMap.mapValues(
        _.groupBy(l, e => e[0]),
        l => l.map(e => e[1]))
)

export async function get_subjects_(dn2opts: DnsOpts) : Promise<Subjects> {
    const dns = hMyMap.keys(dn2opts);

    return await get_subjects(dns, dn2opts, undefined, undefined)
}

export async function get_subjects(dns: Dn[], dn2opts: DnsOpts, search_token: Option<string>, sizelimit: Option<number>) : Promise<Subjects> {
    const parent_dn_to_rdns = fromPairsGrouped(_.compact(dns.map(dn => (
        dn_to_rdn_and_parent_dn(dn)?.oMap(([rdn, parent_dn]) => [parent_dn, rdn])
    ))))

    const r = {};

    await hMyMap.eachAsync(parent_dn_to_rdns, async (rdns, parent_dn) => {
        const sscfg = dn_to_subject_source_cfg(parent_dn)
        if (sscfg) {
            let count = 0;
            for (const rdns_ of _.chunk(rdns, 10)) {
                const subjects = await get_subjects_from_same_branch(sscfg, parent_dn, rdns_, dn2opts, search_token);
                count += _.size(subjects);
                Object.assign(r, subjects)
                if (sizelimit) {
                    if (count >= sizelimit) break;
                }
            }
        }
    })

    return r
}
